packages/dashboard/src/testing/testing-ground/testing-ground.tsx

```diff
@@ -9,6 +17,42 @@ import { dashboardConfig } from '../mocks';
 export class TestingGround {
   @State() dashboardConfiguration: DashboardConfiguration = dashboardConfig;
 
+  /** Holds all necessary information about dashboard */
+  @State() state: DashboardState = {
+    dashboardConfiguration: dashboardConfig,
+    selectedWidgetIds: [],
+    numTimesCopyGroupHasBeenPasted: 0,
+    copyGroup: [],
+    stretchToFit: DEFAULT_STRETCH_TO_FIT,
+    width: DEFAULT_WIDTH,
+    cellSize: DEFAULT_CELL_SIZE,
+    intermediateDashboardConfiguration: undefined,
+    undoQueue: [],
+    redoQueue: [],
+    previousPosition: undefined,
+  };
+
+  /**
+   * Callback that is fired every time the dashboard configuration has been altered.
+   *
+   * When a widget is moved, resized, deleted, appended, or altered, then this method is called
+   */
+  onDashboardConfigurationChange = (config: DashboardConfiguration) => {
```

Comment:
with this 'new iteration' of testing ground, we should instead take a property `onDashboardConfigurationChange` and simply pass down to the child component that expects it

Reply:
I'm a little confused by this request. Can you expand on this a bit? What exactly would this accomplish, and how does this replace internal parts of iot-dashboard-grid / iot-dashboard-grid-internal / iot-annotations dispatching to the store?

Right now (despite the descriptive comment above the method, which is not accurate) this is only being called on initial load of the component, and updates are made by this.store.dispatch() calls. I have not changed anything about this, merely moved it up one component level.

Reply:
I'm not sure what you're asking here exactly.

onDashboardConfigurationChange is a function that we only call inside store.subscribe that merely copies the updated Redux state's dashboardConfiguration into this Stencil component's local dashboardConfiguration state. We only want to call it within store.subscribe, assuming we always want the dashboardConfiguration we are passing to the UI to reflect what it is in Redux.

Are you saying child components of testing-ground should each be doing store.subscribe independently, rather than testing-ground being subscribed to Redux and passing state down? In that case, each child should have its own version of onDashboardConfigurationChange that updates its own internal copy of the Redux state; it doesn't make sense to have them both update their parent's copy of the Redux state when the Redux state changes - we'd be updating it twice on each change.

Reply:
In general, what's happening with state in iot-dashboard currently needs a rethink. We are storing dashboardConfiguration in two different places (the dashboardConfiguration prop and state.dashboardConfiguration), and we are overwriting the dashboardConfiguration prop (!) with the Redux state's dashboardConfiguration when Redux updates.

This is one reason I really don't think we should get too deep into the weeds on trying to fix state on this branch, beyond minimally moving up what needs to move up -- it needs a larger rethink.